[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66967



Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

huiguoo committed:
[tensorexpr] Memory planning: add pre-allocation for reused buffers

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

t10-13rocket reviewed (approved):
LGTM

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

ZolotukhinM commented:
Do you mind elaborate on what this PR is doing (ideally, in the commit message)? Currently it's somewhat unclear: we've already had pre-allocation and reusing for intermediate buffers, what are we adding/changing here and why?

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.h:
```diff
@@ -111,6 +119,7 @@ class TORCH_API CodeGen {
   std::vector<BufferArg> buffer_args_;
   at::Device device_ = at::kCPU;
   std::string kernel_func_name_ = "func";
+  std::vector<BufPtr> buffers_to_alloc_;
```
The names are getting quite confusing now. If one reads this code after this PR lands, they would see `buffer_to_alloc`, `buf_allocs`, `interm_bufs`, `buf_args` and many of those things, despite appearing similar, mean very different things. Let's rename them to better reflect what they hold. E.g. one of them can be `preallocated_bufs` - i.e. a set of list of buffers that don't have Allocate/Free statements and need to be preallocated outside of the codegen (i.e. in the Kernel). The other one can be `dynamically_allocated_bufs` - those are the intermediate bufs that have Allocate/Free/Alias stmt for them. You could choose different names if you'd like but please try to assess the code from a position of a person reading it for the first time.

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -332,13 +332,14 @@ LLVMCodeGen::LLVMCodeGen(
+    : CodeGen(stmt, args, device, kernel_func_name, pre_alloc) {
   impl_ = std::make_unique<LLVMCodeGenImpl>(
-      this->stmt(), args, device, dtype, triple, cpu, attrs);
+      this->stmt(), this->buffer_args(), device, dtype, triple, cpu, attrs);
```
Why is this change needed?

huiguoo replied:
Yes it's needed. The buffer args will be updated in the constructor of its parent (CodeGen).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -185,34 +186,45 @@ std::vector<std::pair<BufPtr, BufPtr>> AllocBufsWithMemReuse(
       BufPtr buf = rit->first;
-      b->prepend_stmt(alloc<Allocate>(buf));
-      b->append_stmt(alloc<Free>(buf));
+      if (pre_alloc) {
+        // Save buffers for buffer preallocation in TEK.
+        buffers_to_pre_alloc_.emplace_back(buf);
+        // Add preallocated buffers to args.
+        buffer_args_.emplace_back(BufHandle(buf));
```
It seems like if `pre_alloc` is true, but the buffer has dynamic shapes, it won't be allocated at all. Here we would expect TEK to pre-allocate it, but TEK would reject to do that since shapes are not constants. Am I missing something? Do you mind adding a test to exercise this?

huiguoo replied:
That's a good point! Have fixed it and added a test.

huiguoo commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -185,34 +186,45 @@ std::vector<std::pair<BufPtr, BufPtr>> AllocBufsWithMemReuse(
 // buffer. In current implementation, we use linear scan for memory reuses.
 // TODO: try more memory reuse algorithms and compare their memory efficiency.
-void CodeGen::allocIntermediateBufs() {
+void CodeGen::allocIntermediateBufs(const bool pre_alloc) {
```
Yeah. We can keep it for better coding though. The future reviewers would know it's read only.

huiguoo committed:
Update on "[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers"




Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1335,17 +1328,18 @@ std::vector<BufPtr> TensorExprKernel::preAllocIntermediateBufs(
     if (!bp) {
-      remaining_interm_bufs.push_back(buf);
-      continue;
+      throw std::runtime_error(
+          "TensorExprKernel: memory allocation failed during preallocation!");
```
Let's use one of TE exceptions instead of generic runtime_error here (see https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/exceptions.h).

huiguoo commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1585,7 +1579,13 @@ void TensorExprKernel::compile() {
+      kernel_func_name_,
+      pre_alloc_);
+
+  if (pre_alloc_) {
```
I understand. This is what we need to go through to build good things. Thanks for letting me know. Otherwise it would take more time and effort for me to maintain this PR.

huiguoo commented:
Most of the changes in this PR are not useful anymore. Closing the PR.
